### DISCUSSION TITLE
Is there some way to fix these types without casting?

### DISCUSSION THREAD

**Discussion by codingismy11to7** [RebeccaStevens/deepmerge-ts#489] [Q&A]
Just picked up this library yesterday, and am trying to figure out why my code doesn't work without casting. I'm doing what I think would be a pretty standard thing for this lib, trying to use it for PATCHes.

i have a stored type and a patch type like (simplified)
```typescript
type Stored = Readonly<{field1: ReqObj, field2?: OptObj}>;
type UpdateStored = Readonly<{field1?: ReqObj, field2?: OptObj}>;
```
where the update fields are all optional and the stored fields are a mixture of required and optional.

i'm using deepmerge in a patch function like
```typescript
const DefaultStored: Stored = {...};

const updateStored = (patch: (old?: Stored) => Stored) => {...};

const doUpdate = (update: UpdateStored) => 
  updateStored(old => deepmerge(old ?? DefaultStored, update));
```

the code and functionality work fine, but unless i cast the `update` parameter or `deepmerge` return with `as Stored`, typescript complains that I'm returning an `UpdateStored` from the `deepmerge` call and refuses to compile. so right now i'm doing

```typescript
const doUpdate = (update: UpdateStored) => 
  updateStored(old => deepmerge(old ?? DefaultStored, update) as Stored);
```

which compiles and works but makes me uneasy. just wondering if anybody else has run into anything like this or if i'm doing something obviously wrong

**Comment by RebeccaStevens** [comment]
Would this be a good representation of what you are trying to do?

```ts
import { deepmerge } from "deepmerge-ts";

type Stored = Readonly<{ field1: string; field2?: string }>;
type UpdateStored = Readonly<{ field1?: string; field2?: string }>;

const DefaultStored: Stored = {
  field1: "a"
};

const updateStored = (patch: (old?: Stored) => Stored) => patch({
  field1: "b"
});

const doUpdate = (update: UpdateStored) =>
  updateStored((old) => deepmerge(old ?? DefaultStored, update));

const result = doUpdate({
  field2: "c"
});
```

Because if so, I can't reproduce your issue. `as Stored` isn't need above.
